File: SafirGuard_Python/safir_guard/quarantine/quarantine_vault.py
import os
import json
import time
import shutil
import hashlib
from typing import List, Dict, Optional

VAULT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "quarantine_vault"))
META_FILE = os.path.join(VAULT_DIR, "quarantine_manifest.json")
XOR_KEY = 0x5A  # Güvenli izolasyon için byte bazlı XOR maskeleme

class QuarantineVault:
    def __init__(self, vault_dir: str = VAULT_DIR):
        self.vault_dir = vault_dir
        self.meta_file = os.path.join(self.vault_dir, "quarantine_manifest.json")
        self._ensure_vault()

    def _ensure_vault(self):
        if not os.path.exists(self.vault_dir):
            os.makedirs(self.vault_dir, exist_ok=True)
        if not os.path.exists(self.meta_file):
            self._save_manifest({})

    def _load_manifest(self) -> Dict[str, dict]:
        try:
            if os.path.exists(self.meta_file):
                with open(self.meta_file, "r", encoding="utf-8") as f:
                    return json.load(f)
        except Exception:
            pass
        return {}

    def _save_manifest(self, manifest: Dict[str, dict]):
        with open(self.meta_file, "w", encoding="utf-8") as f:
            json.dump(manifest, f, indent=2, ensure_ascii=False)

    def _xor_transform(self, data: bytes) -> bytes:
        return bytes([b ^ XOR_KEY for b in data])

    def quarantine_file(self, file_path: str, threat_name: str, threat_type: str = "Malware") -> Optional[dict]:
        if not os.path.isfile(file_path):
            return None

        try:
            with open(file_path, "rb") as f:
                content = f.read()

            sha256 = hashlib.sha256(content).hexdigest()
            file_size = len(content)
            file_name = os.path.basename(file_path)
            q_id = f"Q_{int(time.time())}_{sha256[:8]}"
            vault_filename = f"{q_id}.safir_locked"
            vault_path = os.path.join(self.vault_dir, vault_filename)

            # Dosyayı XOR ile kilitle ve kasaya yaz
            locked_content = self._xor_transform(content)
            with open(vault_path, "wb") as f:
                f.write(locked_content)

            # Orijinal dosyayı güvenle sil
            try:
                os.remove(file_path)
            except Exception as e:
                # Eğer dosya kullanımda ise sıfırla
                pass

            manifest = self._load_manifest()
            entry = {
                "id": q_id,
                "file_name": file_name,
                "original_path": os.path.abspath(file_path),
                "threat_name": threat_name,
                "threat_type": threat_type,
                "file_size": file_size,
                "sha256": sha256,
                "vault_file": vault_filename,
                "quarantine_time": time.strftime("%Y-%m-%d %H:%M:%S")
            }
            manifest[q_id] = entry
            self._save_manifest(manifest)
            return entry

        except Exception as e:
            print(f"[Quarantine Error] {file_path} karantinaya alınamadı: {e}")
            return None

    def restore_file(self, q_id: str) -> bool:
        manifest = self._load_manifest()
        if q_id not in manifest:
            return False

        entry = manifest[q_id]
        vault_path = os.path.join(self.vault_dir, entry["vault_file"])
        orig_path = entry["original_path"]

        if not os.path.exists(vault_path):
            return False

        try:
            with open(vault_path, "rb") as f:
                locked_content = f.read()

            orig_content = self._xor_transform(locked_content)

            orig_dir = os.path.dirname(orig_path)
            if not os.path.exists(orig_dir):
                os.makedirs(orig_dir, exist_ok=True)

            with open(orig_path, "wb") as f:
                f.write(orig_content)

            # Kasadaki kilitli dosyayı ve kaydı temizle
            os.remove(vault_path)
            del manifest[q_id]
            self._save_manifest(manifest)
            return True
        except Exception as e:
            print(f"[Restore Error] {q_id} geri yüklenemedi: {e}")
            return False

    def delete_permanently(self, q_id: str) -> bool:
        manifest = self._load_manifest()
        if q_id not in manifest:
            return False

        entry = manifest[q_id]
        vault_path = os.path.join(self.vault_dir, entry["vault_file"])

        if os.path.exists(vault_path):
            try:
                # Güvenli shred (üzerine yazıp silme)
                size = os.path.getsize(vault_path)
                with open(vault_path, "wb") as f:
                    f.write(b"\x00" * size)
                os.remove(vault_path)
            except Exception:
                pass

        del manifest[q_id]
        self._save_manifest(manifest)
        return True

    def list_quarantined(self) -> List[dict]:
        manifest = self._load_manifest()
        return list(manifest.values())

File: SafirGuard_CSharp/src/SafirGuard.App/Quarantine/QuarantineVault.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using System.Collections.Generic;
using SafirGuard.App.Models;

namespace SafirGuard.App.Quarantine
{
    /// <summary>
    /// Safir Karantina Kasası — Tehditleri Şifreli Olarak İzole Eder
    /// </summary>
    public class QuarantineVault
    {
        private readonly string _vaultDir;
        private readonly string _metaFile;
        private const byte XorKey = 0x5A;
        private readonly object _lock = new();

        public QuarantineVault(string? vaultDir = null)
        {
            _vaultDir = vaultDir ?? Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "quarantine_vault");
            _metaFile = Path.Combine(_vaultDir, "quarantine_manifest.json");
            EnsureVault();
        }

        private void EnsureVault()
        {
            if (!Directory.Exists(_vaultDir))
            {
                Directory.CreateDirectory(_vaultDir);
            }
            if (!File.Exists(_metaFile))
            {
                SaveManifest(new Dictionary<string, QuarantineItem>());
            }
        }

        private Dictionary<string, QuarantineItem> LoadManifest()
        {
            lock (_lock)
            {
                try
                {
                    if (File.Exists(_metaFile))
                    {
                        string json = File.ReadAllText(_metaFile);
                        return JsonSerializer.Deserialize<Dictionary<string, QuarantineItem>>(json) ?? new();
                    }
                }
                catch { }
                return new();
            }
        }

        private void SaveManifest(Dictionary<string, QuarantineItem> manifest)
        {
            lock (_lock)
            {
                string json = JsonSerializer.Serialize(manifest, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(_metaFile, json);
            }
        }

        private byte[] XorTransform(byte[] data)
        {
            byte[] output = new byte[data.Length];
            for (int i = 0; i < data.Length; i++)
            {
                output[i] = (byte)(data[i] ^ XorKey);
            }
            return output;
        }

        public QuarantineItem? QuarantineFile(string filePath, string threatName, string threatType = "Malware")
        {
            if (!File.Exists(filePath)) return null;

            try
            {
                byte[] rawBytes = File.ReadAllBytes(filePath);
                string sha256 = Convert.ToHexString(SHA256.HashData(rawBytes)).ToLowerInvariant();
                string fileName = Path.GetFileName(filePath);
                string qId = $"Q_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_{sha256.Substring(0, 8)}";
                string vaultFileName = $"{qId}.safir_locked";
                string vaultFilePath = Path.Combine(_vaultDir, vaultFileName);

                // XOR şifreleme ve kasaya yazma
                byte[] lockedBytes = XorTransform(rawBytes);
                File.WriteAllBytes(vaultFilePath, lockedBytes);

                // Orijinal dosyayı sil
                try
                {
                    File.Delete(filePath);
                }
                catch { }

                var item = new QuarantineItem
                {
                    Id = qId,
                    FileName = fileName,
                    OriginalPath = Path.GetFullPath(filePath),
                    ThreatName = threatName,
                    ThreatType = threatType,
                    QuarantinedAt = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss"),
                    FileSize = rawBytes.Length,
                    Sha256 = sha256,
                    VaultFile = vaultFileName
                };

                var manifest = LoadManifest();
                manifest[qId] = item;
                SaveManifest(manifest);

                return item;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[QuarantineVault Error] {ex.Message}");
                return null;
            }
        }

        public bool RestoreFile(string qId)
        {
            var manifest = LoadManifest();
            if (!manifest.TryGetValue(qId, out var item)) return false;

            string vaultFilePath = Path.Combine(_vaultDir, item.VaultFile);
            if (!File.Exists(vaultFilePath)) return false;

            try
            {
                byte[] lockedBytes = File.ReadAllBytes(vaultFilePath);
                byte[] origBytes = XorTransform(lockedBytes);

                string? origDir = Path.GetDirectoryName(item.OriginalPath);
                if (!string.IsNullOrEmpty(origDir) && !Directory.Exists(origDir))
                {
                    Directory.CreateDirectory(origDir);
                }

                File.WriteAllBytes(item.OriginalPath, origBytes);
                File.Delete(vaultFilePath);

                manifest.Remove(qId);
                SaveManifest(manifest);
                return true;
            }
            catch
            {
                return false;
            }
        }

        public bool DeletePermanently(string qId)
        {
            var manifest = LoadManifest();
            if (!manifest.TryGetValue(qId, out var item)) return false;

            string vaultFilePath = Path.Combine(_vaultDir, item.VaultFile);
            if (File.Exists(vaultFilePath))
            {
                try
                {
                    // Güvenli silme: dosyanın üzerine sıfır yaz
                    long len = new FileInfo(vaultFilePath).Length;
                    File.WriteAllBytes(vaultFilePath, new byte[len]);
                    File.Delete(vaultFilePath);
                }
                catch { }
            }

            manifest.Remove(qId);
            SaveManifest(manifest);
            return true;
        }

        public List<QuarantineItem> ListQuarantined()
        {
            return new List<QuarantineItem>(LoadManifest().Values);
        }
    }
}
